### PR TITLE
SP-177 Google Places autocomplete middleware

### PR DESCRIPTION
**Pull request by joshuapaling**
JIRA: https://aussiecommerce.atlassian.net/browse/SP-177

This adds a proxy for the Google Places autocomplete endpoint. It doesn't change the response at all yet - but down the track we'll eg. push "Bali" up to the top when users search for "Bal".

This PR also adds some error handling and does some test and types tidy up. Detailed commits tell the story. Down the bottom are screenshots of Sentry - it looks pretty good for monitoring exceptions.

ToDo: in future PR, note that google place IDs *can* change: https://developers.google.com/places/web-service/place-id. This will be relevant when we boost specific place IDs like Bali.

Screen Shot 2020-08-12 at 9 06 53 am
Screen Shot 2020-08-12 at 9 07 37 am

### CODE REVIEW

File: src/lib/address.ts [hunk 1]
```diff
@@ -32,27 +33,30 @@ function buildAlertFromComponents(
...
-  const location = address.geometry.location;
+  const addressComponents = address.address_components || [];
+  const location = address.geometry?.location;
   return {
     brand: locationAlert.brand,
     place_id: locationAlert.place_id,
     google_result: {
       continent: "",
-      country: getValue("country", addressComponents),
+      country: getValue(GM.AddressType.country, addressComponents),
```
Comment (shawn-cx-li) [inline src/lib/address.ts:43]: 👍

File: src/server.ts [hunk 2]
```diff
@@ -7,11 +9,21 @@ import * as adminController from "./controllers/adminController";
...
 import * as popularDestinationController from "./controllers/popularDestinationController";
+import * as googlePlaceController from "./controllers/googlePlaceController";
 import { authMiddleware, adminMiddleware } from "./middleware/auth";
 import { corsMiddleware } from "./middleware/cors";
 
 export default function server(): any {
   const app: express.Application = express();
+  Sentry.init({
+    dsn:
+      "https://d030249111a144c29902b24d4ea41a20@o273699.ingest.sentry.io/5385410"
```
Comment (shawn-cx-li) [inline src/server.ts:20]: can we move `dsn` to env or config file?

File: src/lib/address.ts [hunk 3]
```diff
@@ -32,27 +33,30 @@ function buildAlertFromComponents(
...
+      )
     },
     location_alert: {
       level: locationAlert.location_alert.level,
       value: getAlertLevelValue(locationAlert, addressComponents),
       geocode: {
-        lat: location.lat,
-        lng: location.lng
+        lat: location?.lat,
+        lng: location?.lng
```
Comment (colins44) [inline src/lib/address.ts:59]: I have read about this syntax but never used it, good highlight, I will start using this from now on.

File: src/server.ts [hunk 4]
```diff
@@ -7,11 +9,21 @@ import * as adminController from "./controllers/adminController";
...
+  });
+  app.use(
+    Sentry.Handlers.requestHandler({
+      user: ["email", "roles", "SubscriberKey", "herokuId"]
```
Comment (colins44) [inline src/server.ts:24]: I tend to stay away from sending email address around systems when I can, even if that is to trusted third party services. 

Would there be a reason that you would need an email within sentry?

Comment (joshuapaling) [reply src/server.ts]: Reason is it's convenient. But you're right, not worth the risk, and we can use the `herokuId` to look up the email anyway. I've amended the commit to remove email.

**Pull request merged**